### PR TITLE
B2CA-1247: Fix deprecated warnings

### PR DESCRIPTION
**Pull request by cedelavergne-ledger**
This PR Adds the Guideline enforcer CI.

In order to have those workflows functionnal and :green_circle:, this PR also contains:

- Fix the compilation warnings (like _deprecated_)
- Fix `scan-build` issues
- Fix mispelling and various linter errors
- Adapt `ci-workflow` to reusable and guideline enforrcer
- Fix glyphs and icons
- Update version to **1.11.1**

## Changes include

- [X] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Tests
- [ ] Documentation
- [ ] Other (for changes that might not fit in any category)

### CODE REVIEW

**Review by fbeutin-ledger — COMMENTED** [review]
Small question, why remove .github/workflows/auto-author-assign.yml ?
(No opinion either way, just curious)

**Comment by cedelavergne-ledger** [comment]
> Small question, why remove .github/workflows/auto-author-assign.yml ? (No opinion either way, just curious)

Because just assigning the author to his own PR seems not usefull.

File: src/handle_check_address.c [hunk 1]
```diff
@@ -18,76 +19,45 @@ void handle_check_address(check_address_parameters_t* params, chain_config_t* ch
...
-        char address[51];
-    } locals_union1;
-    union group2 {
-        uint8_t privateKeyData[64];
-        cx_ecfp_public_key_t publicKey;
-    } locals_union2;
+    uint32_t bip32Path[MAX_BIP32_PATH];
+    char address[51];
+    cx_ecfp_256_private_key_t priv;
+    cx_ecfp_256_public_key_t pub;
```
Comment (fbeutin-ledger) [inline src/handle_check_address.c:25]: IIRC we had to do the union optimization or we had stack issues :/

File: src_features/signMessageEIP712_common/common_712.c [hunk 2]
```diff
@@ -8,62 +9,52 @@
...
+        THROW(0x6F00);
+    }
     G_io_apdu_buffer[0] = 27;
     if (info & CX_ECCINFO_PARITY_ODD) {
         G_io_apdu_buffer[0]++;
     }
     if (info & CX_ECCINFO_xGTn) {
         G_io_apdu_buffer[0] += 2;
     }
-    format_signature_out(signature);
```
Comment (fbeutin-ledger) [inline src_features/signMessageEIP712_common/common_712.c:66]: this did nothing ??

File: src_features/signMessage/ui_common_signMessage.c [hunk 3]
```diff
@@ -1,39 +1,29 @@
...
+        THROW(0x6F00);
+    }
     G_io_apdu_buffer[0] = 27;
     if (info & CX_ECCINFO_PARITY_ODD) {
         G_io_apdu_buffer[0]++;
     }
     if (info & CX_ECCINFO_xGTn) {
         G_io_apdu_buffer[0] += 2;
     }
-    format_signature_out(signature);
```
Comment (fbeutin-ledger) [inline src_features/signMessage/ui_common_signMessage.c:36]: same.. weird dead code

File: src/handle_check_address.c (hunk 1 above)
Comment (cedelavergne-ledger) [reply src/handle_check_address.c:25]: After using `crypto_helpers.h`, we don't need exactly the same parameters.
Before the change, we had 2 `union` groups, but now, due with the new function calls, only `address[51]` can be mixed with `eitherbip32Path` or `priv` parameters. All other combinations are impossible because the parameters are used simultaeously.
Considering the sizes, we could use:
```
    union group {
        cx_ecfp_256_private_key_t priv;
        char address[51];
    } locals_union;
```

Do you think it is absolutely necessary?

Also, I have modified `cx_ecfp_private_key_t`/`cx_ecfp_public_key_t` by `cx_ecfp_256_private_key_t`/`cx_ecfp_256_public_key_t`, because it was already done like this in previous porting tasks. Which do you think is better (they are the same today)? Personnally, I prefer the version `xx_256_xx`, which is more explicit...

File: src/handle_check_address.c [hunk 4]
```diff
@@ -18,76 +19,45 @@ void handle_check_address(check_address_parameters_t* params, chain_config_t* ch
...
     }
 
-    if (strcmp(locals_union1.address, params->address_to_check + offset_0x) != 0) {
+    if (strcmp(address, params->address_to_check + offset_0x) != 0) {
         PRINTF("Addresses don't match\n");
     } else {
         PRINTF("Addresses match\n");
         params->result = 1;
     }
-    ZERO(locals_union1);
```
[the remark this answers is not in the document]
Comment (cedelavergne-ledger) [reply src/handle_check_address.c:92]: Ok, yes. Done!

File: src_features/signMessage/ui_common_signMessage.c (hunk 3 above)
Comment (cedelavergne-ledger) [reply src_features/signMessage/ui_common_signMessage.c:36]: same answer

File: src/handle_check_address.c (hunk 1 above)
Comment (fbeutin-ledger) [reply src/handle_check_address.c]: > Do you think it is absolutely necessary?

Mmh this should be ok, address handling is not the most ram intensive task I think

cx_ecfp_256_private_key_t/cx_ecfp_256_public_key_t is fine 👍

File: src_common/common_utils.h [hunk 5]
```diff
@@ -35,6 +35,10 @@ static const char HEXDIGITS[] = "0123456789abcdef";
 
 #define ARRAY_SIZE(a) (sizeof(a) / sizeof(a[0]))
 
+extern size_t strlcpy(char *dst, const char *src, size_t size);
+
+extern size_t strlcat(char *dst, const char *src, size_t size);
```
Comment (fbeutin-ledger) [inline src_common/common_utils.h:40]: Why is this needed ?
I think this will cause an issue for plugins, the `src_common` dir is shared with them

File: src_common/common_utils.c [hunk 6]
```diff
@@ -214,50 +214,25 @@ bool amountToString(const uint8_t *amount,
     return true;
 }
 
-bool getEthAddressFromKey(cx_ecfp_public_key_t *publicKey, uint8_t *out, cx_sha3_t *sha3Context) {
+void getEthAddressFromRawKey(const uint8_t raw_pubkey[static 65],
```
Comment (fbeutin-ledger) [inline src_common/common_utils.c:217]: I think I prefer the old version tbh
Plus this is used by plugins and we don't really want to break the compatibility

File: src_common/common_utils.h [hunk 7]
```diff
@@ -56,12 +60,14 @@ bool adjustDecimals(const char *src,
...
+                             cx_sha3_t *sha3Context);
 
-bool getEthAddressStringFromKey(cx_ecfp_public_key_t *publicKey,
-                                char *out,
-                                cx_sha3_t *sha3Context,
-                                uint64_t chainId);
+void getEthAddressStringFromRawKey(const uint8_t raw_pubkey[static 65],
+                                   char *out,
+                                   cx_sha3_t *sha3Context,
+                                   uint64_t chainId);
```
Comment (fbeutin-ledger) [inline src_common/common_utils.h:70]: same

File: src_features/getPublicKey/cmd_getPublicKey.c [hunk 8]
```diff
@@ -35,25 +34,19 @@ void handleGetPublicKey(uint8_t p1,
...
-                                    chainConfig->chainId)) {
-        THROW(CX_INVALID_PARAMETER);
+    if (bip32_derive_get_pubkey_256(
+            CX_CURVE_256K1,
+            bip32.path,
+            bip32.length,
+            tmpCtx.publicKeyContext.publicKey.W,
+            (tmpCtx.publicKeyContext.getChaincode ? tmpCtx.publicKeyContext.chainCode : NULL),
+            CX_SHA512) != CX_OK) {
+        THROW(0x6F00);
```
Comment (fbeutin-ledger) [inline src_features/getPublicKey/cmd_getPublicKey.c:44]: do we have a defined apdu for this ?

File: src_common/common_utils.h (hunk 5 above)
Comment (cedelavergne-ledger) [reply src_common/common_utils.h:40]: Without such, I still have a compilation warning, because the function declaration is unknown... :thinking: 
Do you see another way to declare them? There are unknown in the sdk either...

File: src_features/getPublicKey/cmd_getPublicKey.c (hunk 8 above)
Comment (cedelavergne-ledger) [reply src_features/getPublicKey/cmd_getPublicKey.c:44]: What do you mean, instead of the error code 0x6F00? I could rather use a `CX_ASSERT`?

File: src_features/provideErc20TokenInformation/cmd_provideTokenInfo.c [hunk 9]
```diff
@@ -168,45 +168,21 @@ void handleProvideErc20TokenInformation(uint8_t p1,
...
-                             workBuffer + offset,
-                             dataLength)) {
+    } else
+#endif
+    {
+        CX_ASSERT(cx_ecfp_init_public_key_no_throw(CX_CURVE_256K1,
+                                                   LEDGER_SIGNATURE_PUBLIC_KEY,
+                                                   sizeof(LEDGER_SIGNATURE_PUBLIC_KEY),
+                                                   &tokenKey));
+        if (!cx_ecdsa_verify_no_throw(&tokenKey, hash, 32, workBuffer + offset, dataLength)) {
```
Comment (fbeutin-ledger) [inline src_features/provideErc20TokenInformation/cmd_provideTokenInfo.c:178]: A bit hard to read with all the ifdef
can you please add some comments to help ?

File: src_features/provideNFTInformation/cmd_provideNFTInfo.c [hunk 10]
```diff
@@ -159,20 +159,10 @@ void handleProvideNFTInformation(uint8_t p1,
...
-            hashId = CX_SHA256;
-            break;
-        default:
-            PRINTF("Incorrect algorithmId %d\n", algorithmId);
-            THROW(0x6a80);
-            break;
+
+    if (algorithmId != ALGORITHM_ID_1) {
+        PRINTF("Incorrect algorithmId %d\n", algorithmId);
+        THROW(0x6a80);
```
Comment (fbeutin-ledger) [inline src_features/provideNFTInformation/cmd_provideNFTInfo.c:165]: do we have a defined apdu code for this ?

File: src_features/signMessage/ui_common_signMessage.c [hunk 11]
```diff
@@ -1,39 +1,29 @@
...
+                                            tmpCtx.messageSigningContext.bip32.path,
+                                            tmpCtx.messageSigningContext.bip32.length,
+                                            CX_RND_RFC6979 | CX_LAST,
+                                            CX_SHA256,
+                                            tmpCtx.messageSigningContext.hash,
+                                            sizeof(tmpCtx.messageSigningContext.hash),
+                                            G_io_apdu_buffer + 1,
+                                            G_io_apdu_buffer + 1 + 32,
+                                            &info) != CX_OK) {
+        THROW(0x6F00);
```
Comment (fbeutin-ledger) [inline src_features/signMessage/ui_common_signMessage.c:18]: defined error code ?

File: .gitignore [hunk 12]
```diff
@@ -8,6 +8,7 @@ obj/
 build/
```
Comment (xchapron-ledger) [inline .gitignore:8]: You can remove most of the above and only keep bin, debug and build.

File: Makefile
Comment (xchapron-ledger) [inline Makefile:1]: Would be very nice to use makefile.standard_app but maybe you keep it for a latter Pr?

File: src_features/signMessageEIP712_common/common_712.c [hunk 13]
```diff
@@ -8,62 +9,52 @@
...
+                                            tmpCtx.messageSigningContext712.bip32.path,
+                                            tmpCtx.messageSigningContext712.bip32.length,
+                                            CX_RND_RFC6979 | CX_LAST,
+                                            CX_SHA256,
+                                            hash,
+                                            sizeof(hash),
+                                            G_io_apdu_buffer + 1,
+                                            G_io_apdu_buffer + 1 + 32,
+                                            &info) != CX_OK) {
+        THROW(0x6F00);
```
Comment (fbeutin-ledger) [inline src_features/signMessageEIP712_common/common_712.c:49]: eplicit error code ?

File: src/handle_check_address.c [hunk 14]
```diff
@@ -18,76 +19,47 @@ void handle_check_address(check_address_parameters_t* params, chain_config_t* ch
...
-    if (os_derive_bip32_no_throw(CX_CURVE_256K1,
-                                 locals_union1.bip32Path,
-                                 bip32PathLength,
-                                 locals_union2.privateKeyData,
-                                 NULL) != CX_OK) {
-        ZERO(locals_union1);
-        ZERO(locals_union2);
-        return;
-    }
+    CX_CHECK(bip32_derive_with_seed_init_privkey_256(HDW_NORMAL,
```
Comment (xchapron-ledger) [inline src/handle_check_address.c:37]: Any reason not to sue bip32_derive_get_pubkey_256?

File: src_nbgl/ui_sign_712.c [hunk 15]
```diff
@@ -48,18 +48,10 @@ static bool display_review_page(uint8_t page, nbgl_pageContent_t *content) {
             break;
 
         case 1:
-            switch (ui_712_next_field()) {
-                case EIP712_NO_MORE_FIELD:
-                    ui_712_switch_to_sign();
-                    ret = true;
```
Comment (fbeutin-ledger) [inline src_nbgl/ui_sign_712.c:54]: This `ret == true` was never used right?
Your code is equivalent but I think there is a deeper problem here..

File: src/main.c [hunk 16]
```diff
@@ -162,11 +141,17 @@ void handleGetWalletId(volatile unsigned int *tx) {
     unsigned char t[64];
```
Comment (xchapron-ledger) [inline src/main.c:141]: IMO HAVE_WALLET_ID_SDK should be dropped. See https://ledgerhq.atlassian.net/browse/EMBSDK-18

File: tests/speculos/test_configuration_cmd.py [hunk 17]
```diff
@@ -1,3 +1,3 @@
 
 def test_configuration(cmd):
-    assert cmd.get_configuration() == (2, 1, 11, 0)
+    assert cmd.get_configuration() == (2, 1, 11, 1)
```
Comment (fbeutin-ledger) [inline tests/speculos/test_configuration_cmd.py:3]: You can use something like here https://github.com/LedgerHQ/app-exchange/blob/41d158b4e7eb54f78c2ef5f6f4a1c93819a64f63/test/python/utils.py#L29

File: src_common/common_utils.c [hunk 18]
```diff
@@ -214,50 +214,25 @@ bool amountToString(const uint8_t *amount,
...
-                         publicKey->W + 1,
-                         64,
-                         hashAddress,
-                         32) != CX_OK) {
-        return false;
-    }
-
+    CX_ASSERT(cx_keccak_init_no_throw(sha3Context, 256));
+    CX_ASSERT(
+        cx_hash_no_throw((cx_hash_t *) sha3Context, CX_LAST, raw_pubkey + 1, 64, hashAddress, 32));
```
Comment (xchapron-ledger) [inline src_common/common_utils.c:223]: Should use `cx_keccak_256_hash` to simplify and spare RAM if at some point we changed the prototype (cx_sha3_t will be unused).

File: src_nbgl/ui_sign_712.c (hunk 15 above)
Comment (fbeutin-ledger) [reply src_nbgl/ui_sign_712.c:54]: cc @apaillier-ledger

File: src_common/common_utils.c [hunk 19]
```diff
@@ -214,50 +214,25 @@ bool amountToString(const uint8_t *amount,
     return true;
 }
 
-bool getEthAddressFromKey(cx_ecfp_public_key_t *publicKey, uint8_t *out, cx_sha3_t *sha3Context) {
+void getEthAddressFromRawKey(const uint8_t raw_pubkey[static 65],
+                             uint8_t *out,
```
Comment (xchapron-ledger) [inline src_common/common_utils.c:218]:
```suggestion
                             uint8_t out[static CX_KECCAK_256_SIZE],
```

File: src_common/common_utils.c (hunk 18 above)
Comment (xchapron-ledger) [reply src_common/common_utils.c:223]: Same below

File: src_common/common_utils.c (hunk 19 above)
Comment (xchapron-ledger) [reply src_common/common_utils.c:218]: Same below

File: src_common/common_utils.h (hunk 5 above)
Comment (xchapron-ledger) [reply src_common/common_utils.h:40]: this is part of `string.h` with our toolchain version I think, so should be replace with a #include ?

File: src_features/signMessageEIP712_common/common_712.c (hunk 2 above)
Comment (fbeutin-ledger) [reply src_features/signMessageEIP712_common/common_712.c:66]: cc @apaillier-ledger you might know more ?

Comment (xchapron-ledger) [reply src_features/signMessageEIP712_common/common_712.c:66]: Yeap before we had:
- sign in signature with cx_ecdsa_sign
- extract r and s from signature and put it in G_io_apdu_buffer with format_signature_out

Now both are done in once with a single bip32_derive_ecdsa_sign_rs_hash_256 call.

Note that I didn't see the removal of the `format_signature_out`, can't it be removed now?

Comment (cedelavergne-ledger) [reply src_features/signMessageEIP712_common/common_712.c:66]: `format_signature_out` has been removed, because no more called anywhere.

File: src_features/performPrivacyOperation/cmd_performPrivacyOperation.c [hunk 20]
```diff
@@ -53,27 +53,31 @@ void handlePerformPrivacyOperation(uint8_t p1,
...
-        (tmpCtx.publicKeyContext.getChaincode ? tmpCtx.publicKeyContext.chainCode : NULL));
-    cx_ecfp_init_private_key(CX_CURVE_256K1, privateKeyData, 32, &privateKey);
-    cx_ecfp_generate_pair(CX_CURVE_256K1, &tmpCtx.publicKeyContext.publicKey, &privateKey, 1);
-    if (!getEthAddressStringFromKey(&tmpCtx.publicKeyContext.publicKey,
-                                    tmpCtx.publicKeyContext.address,
-                                    &global_sha3,
-                                    chainConfig->chainId)) {
-        THROW(CX_INVALID_PARAMETER);
-    }
+        (tmpCtx.publicKeyContext.getChaincode ? tmpCtx.publicKeyContext.chainCode : NULL)));
```
[the remark this answers is not in the document]
Comment (cedelavergne-ledger) [reply src_features/performPrivacyOperation/cmd_performPrivacyOperation.c:61]: :thinking: clang-format is happy with that...

File: src_features/provideErc20TokenInformation/cmd_provideTokenInfo.c (hunk 9 above)
Comment (cedelavergne-ledger) [reply src_features/provideErc20TokenInformation/cmd_provideTokenInfo.c:178]: In fact, the calls to `cx_ecfp_init_public_key` and `cx_ecdsa_verify` were done twice, with the same parameters, just because of the `#ifdef` which was encapsulating the `if/else.`

The code was:
```
#ifdef HAVE_TOKENS_EXTRA_LIST
...
if (condition) {
    PRINTF()
} else {
    cx_ecfp_init_public_key()
    cx_ecdsa_verify()
}
#else
cx_ecfp_init_public_key()
cx_ecdsa_verify()
#endif
```

Which I simplified with:
```
#ifdef HAVE_TOKENS_EXTRA_LIST
...
if (condition) {
    PRINTF()
} else
#endif
{
    cx_ecfp_init_public_key()
    cx_ecdsa_verify()
}
```

File: src_features/provideNFTInformation/cmd_provideNFTInfo.c (hunk 10 above)
Comment (cedelavergne-ledger) [reply src_features/provideNFTInformation/cmd_provideNFTInfo.c:165]: Yes, we have this one:
`#define APDU_RESPONSE_INVALID_DATA            0x6a80`

I have just kept the same logic as most of the code in this file... But Ok, I will change the hardcoded values by the define

File: src_features/signMessage/ui_common_signMessage.c (hunk 11 above)
Comment (cedelavergne-ledger) [reply src_features/signMessage/ui_common_signMessage.c:18]: Not yet.
I have added `APDU_RESPONE_UNKNOWN`.

File: src_features/signMessageEIP712_common/common_712.c (hunk 13 above)
Comment (cedelavergne-ledger) [reply src_features/signMessageEIP712_common/common_712.c:49]: Same

File: src_nbgl/ui_sign_712.c (hunk 15 above)
Comment (cedelavergne-ledger) [reply src_nbgl/ui_sign_712.c:54]: The previous code was setting `ret = true`, but just behind, we had `__attribute__((fallthrough))`, which set `ret = false`.
That means the `ret = true` was useless, and I have simplified the code.

File: tests/speculos/test_configuration_cmd.py (hunk 17 above)
Comment (cedelavergne-ledger) [reply tests/speculos/test_configuration_cmd.py:3]: Of course, and this is already planned in another Jira ticket B2CA-901, related to the ragger tests.

File: .gitignore (hunk 12 above)
Comment (cedelavergne-ledger) [reply .gitignore:8]: Done

File: Makefile
Comment (cedelavergne-ledger) [reply Makefile:1]: This is not the goal of this PR, effectively :wink:

File: src/main.c (hunk 16 above)
Comment (cedelavergne-ledger) [reply src/main.c:141]: :thinking: Should be part of a dedicated ticket and PR, I guess?

File: src_common/common_utils.c (hunk 18 above)
Comment (cedelavergne-ledger) [reply src_common/common_utils.c:223]: Done

File: src_common/common_utils.c (hunk 6 above)
Comment (cedelavergne-ledger) [reply src_common/common_utils.c:217]: Is it directly used by the plugins, or indirectly from the source files from src_plugins/xxx? In which case, there is issue??

File: src/handle_check_address.c (hunk 14 above)
Comment (cedelavergne-ledger) [reply src/handle_check_address.c:37]: Done!

File: src_common/common_utils.h (hunk 5 above)
Comment (cedelavergne-ledger) [reply src_common/common_utils.h]: Seems no more needed... I removed these lines.

File: src_features/provideErc20TokenInformation/cmd_provideTokenInfo.c (hunk 9 above)
Comment (fbeutin-ledger) [reply src_features/provideErc20TokenInformation/cmd_provideTokenInfo.c:178]: Yes sorry we misunderstood: I ended up nderstanding what was done, but I think the code would benefit from comments, preferably a paragraph that explains what is done on each case

File: src_nbgl/ui_sign_712.c (hunk 15 above)
Comment (fbeutin-ledger) [reply src_nbgl/ui_sign_712.c:54]: Ok but this is a bug (present previously but still a bug), the function should return `true` if it succeeds

File: src_common/common_utils.c (hunk 6 above)
Comment (fbeutin-ledger) [reply src_common/common_utils.c:217]: At least Opensea
https://github.com/LedgerHQ/app-plugin-opensea/blob/c64c989e4965c9654f43208123a949bbc4b5e922/src/utils.c#L22

Comment (fbeutin-ledger) [reply src_common/common_utils.c:217]: But it's archived 🤔

Comment (fbeutin-ledger) [reply src_common/common_utils.c:217]: Just in case, can you leave old implem behind with deprecate flag ?

Comment (xchapron-ledger) [reply src_common/common_utils.c:217]: Would be nice to be able to build all deployed plugins against a specific ethereum SDK ref.
To be able to easily test if a change will create deployments issues.

File: src_common/common_utils.c [hunk 21]
```diff
@@ -214,55 +214,23 @@ bool amountToString(const uint8_t *amount,
...
-                         CX_LAST,
-                         publicKey->W + 1,
-                         64,
-                         hashAddress,
-                         32) != CX_OK) {
-        return false;
-    }
-
+    CX_ASSERT(cx_keccak_256_hash(raw_pubkey + 1, 64, hashAddress));
     memmove(out, hashAddress + 12, 20);
```
Comment (xchapron-ledger) [inline src_common/common_utils.c:221]:
```suggestion
    memmove(out, hashAddress + 12, ADDRESS_LENGTH);
```
?

Comment (cedelavergne-ledger) [reply src_common/common_utils.c:221]: Fixed!

File: src_nbgl/ui_sign_712.c (hunk 15 above)
Comment (apaillier-ledger) [reply src_nbgl/ui_sign_712.c:54]: Not necessarily a bug since everything is already behaving (functionally speaking) well.
This return value controls if NBGL should refresh the screen if I recall correctly. :thinking:

**Review by fbeutin-ledger — APPROVED** [review]

**Review by fbeutin-ledger — APPROVED** [review]

**Comment by apaillier-ledger** [comment]
The plugin SDK changelog needs to be updated.
